Network/DNS/Resolver.hs

```diff
@@ -342,15 +369,37 @@ lookupRawInternal rcv ad rlv dom typ = do
               Nothing  -> loop query checkSeqno (cnt + 1) False
               Just res -> do
                   let valid = checkSeqno res
-                  if valid then
-                      return $ Right res
-                    else
-                      loop query checkSeqno (cnt + 1) False
+                  case valid of
+                      False  -> loop query checkSeqno (cnt + 1) False
+                      True | False <- trunCation $ flags $ header res
```

Comment:
The code in question is using Haskell 2010 *pattern guards*:

https://wiki.haskell.org/Pattern_guard

Do you dislike pattern guards? I did not want to parse the headers of a message with the wrong sequence number.

Speaking of sequence timers, I think there's a problem with UDP port re-use in the library:

* Suppose a client is using a single Resolver object to make many consecutive DNS queries.
* Further suppose that one of the queries times out
* Suppose also that the server did not actually lose it, and responds shortly after.

We now have a situation where the client and server are of sync!  The server's answer will now be seen by the client in response to a different query.  The re-use of the same sequence number for query retries does not solve the problem.  After the client finally gets the expected answer there are still more answers pending than it has read, and those answers will arrive in response to subsequent queries.

Therefore, either each query needs to always create a new socket that's used only for retries of that query and then closed, or else the socket needs to be closed after any query that encounters a timeout and needs to retransmit.  The always use a new socket strategy seems simpler.

I would therefore propose that "withResolver" largely become a noop that squirrels away the seed, but does not preemptively open the socket, instead a new socket is always opened in `lookupRawInternal` and used only for a single retry loop.  The cleanup `bracket` for the socket would now happen in `lookupRawInternal`.

Thoughts?  I can implement this if you don't have a different solution in mind?